qiskit/circuit/library/fourier_checking.py

```diff
@@ -81,17 +87,72 @@ def __init__(self, f: List[int], g: List[int]) -> None:
                 "{1, -1}."
             )
 
-        circuit = QuantumCircuit(num_qubits, name=f"fc: {f}, {g}")
-
+        # This definition circuit is not replaced by the circuit produced by fourier_checking,
+        # as the latter produces a slightly different circuit, with DiagonalGates instead
+        # of Diagonal circuits.
+        circuit = QuantumCircuit(int(num_qubits), name=f"fc: {f}, {g}")
         circuit.h(circuit.qubits)
-
         circuit.compose(Diagonal(f), inplace=True)
-
         circuit.h(circuit.qubits)
-
         circuit.compose(Diagonal(g), inplace=True)
-
         circuit.h(circuit.qubits)
-
         super().__init__(*circuit.qregs, name=circuit.name)
         self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+
+
+def fourier_checking(f: List[int], g: List[int]) -> QuantumCircuit:
```

Comment:
It would be nice to use `collections.abc.Sequence` in consistency with `DiagonalGate`, but we can also do this later 🙂 
```suggestion
def fourier_checking(f: Sequence[int], g: Sequence[int]) -> QuantumCircuit:
```

Reply:
sure, done in [36e3450](https://github.com/Qiskit/qiskit/pull/13353/commits/36e3450bf2f14eccffd5fcb574e9fb78f2c06d98)